lib/src/suggestor_mixins.dart

```diff
@@ -0,0 +1,114 @@
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:codemod/src/file_context.dart';
+
+import 'patch.dart';
+import 'suggestor.dart';
+
+/// Mixin that implements the [Suggestor] interface and makes it easier to write
+/// suggestors that operate as an [AstVisitor].
+///
+/// With the [AstVisitor] pattern, you can override the applicable `visit`
+/// methods to find what you're looking for and generate patches at specific
+/// locations in the source using the offsets provided by the [AstNode]s and
+/// tokens therein.
+///
+/// Note that this mixin provides an implementation of [generatePatches] that
+/// should not need to be overridden except for performance optimization reasons
+/// like avoiding analysis on certain files.
+///
+/// By default, this operates on the unresolved AST. Subclasses that need a
+/// fully resolved AST (e.g. for static typing info) should override
+/// [shouldResolveAst] to return true.
+///
+/// The easiest way to understand this pattern is to see an example. Consider
+/// the following suggestor that aims to remove all deprecated declarations:
+///
+///     import 'package:analyzer/analyzer.dart';
+///     import 'package:codemod/codemod.dart';
+///
+///     class DeprecatedRemover extends GeneralizingAstVisitor
+///         with UnresolvedAstVisitingSuggestor {
+///
+///       static bool isDeprecated(AnnotatedNode node) =>
+///           node.metadata.any((m) => m.name.name.toLowerCase() == 'deprecated');
+///
+///       @override
+///       visitDeclaration(Declaration node) {
+///         if (isDeprecated(node)) {
+///           // Remove the node by replacing the span from its start offset to its end
+///           // offset with an empty string.
+///           yieldPatch('', node.offset, node.end);
+///         }
+///       }
+///     }
+mixin AstVisitingSuggestor<R> on AstVisitor<R> implements Suggestor {
+  final _patches = <Patch>{};
+
+  /// The context helper for the file for which patches are currently being
+  /// generated.
+  FileContext get context => _context;
+  FileContext _context;
+
+  @override
+  Stream<Patch> generatePatches(FileContext context) async* {
+    _patches.clear();
+    _context = context;
+
+    final unit = shouldResolveAst(context)
+        ? (await context.getResolvedUnit()).unit
+        : context.getUnresolvedUnit();
```

Comment:
Now any suggestor that uses this mixin can easily opt-in to getting the fully resolved AST just by overriding `shouldResolveAst()`. That method can make its determination based on the file contents, if that's helpful, or it can just always return `true`.